internal/pkg/manifest/pipeline.go

```diff
@@ -0,0 +1,204 @@
+		return structs.Map(props), nil
+	}
+	return nil, &ErrMissingProviderProperties{
+		provider: p,
+	}
+}
+
+// PipelineSchemaMajorVersion is the major version number
+// of the pipeline manifest schema
+type PipelineSchemaMajorVersion int
+
+const (
+	// Ver1 is the current schema major version of the pipeline.yml file.
+	Ver1 PipelineSchemaMajorVersion = iota + 1
+)
+
+// PipelineManifest contains information that defines the relationship
+// and deployment ordering of your environments.
+type PipelineManifest struct {
+	Version      PipelineSchemaMajorVersion `yaml:"version"`
+	Source       *Source                    `yaml:"source"`
+	Environments []PipelineStage            `yaml:"stages"`
+}
+
+// Source defines the source of the artifacts to be built and deployed.
+type Source struct {
+	ProviderName string                 `yaml:"provider"`
+	Properties   map[string]interface{} `yaml:"properties"`
+}
+
+// PipelineStage represents configuration for each deployment stage
+// of an application.
+type PipelineStage struct {
+	Name string `yaml:"env"`
+}
+
+// CreatePipeline returns a pipeline manifest object.
+func CreatePipeline(provider Provider, stages ...PipelineStage) (archer.Manifest, error) {
+	// TODO: #221 Do more validations
+	props, err := provider.properties()
+	if err != nil {
+		return nil, err
+	}
+
+	var defaultStages []PipelineStage
+	if len(stages) == 0 {
+		defaultStages = []PipelineStage{
+			{
+				Name: "Test",
+			},
+			{
+				Name: "Prod",
+			},
+		}
+	}
+
+	return &PipelineManifest{
+		Version: Ver1,
+		Source: &Source{
+			ProviderName: provider.String(),
+			Properties:   props,
+		},
+		Environments: append(defaultStages, stages...),
+	}, nil
+}
+
+// Marshal serializes the pipeline manifest object into byte array that
+// represents the pipeline.yml document.
+func (m *PipelineManifest) Marshal() ([]byte, error) {
+	box := templates.Box()
+	content, err := box.FindString("cicd/pipeline.yml")
+	if err != nil {
+		return nil, err
+	}
+	tpl, err := template.New("pipelineTemplate").Parse(content)
+	if err != nil {
+		return nil, err
+	}
+	var buf bytes.Buffer
+	if err := tpl.Execute(&buf, *m); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// UnmarshalPipeline deserializes the YAML input stream into a pipeline
+// manifest object. It returns an error if any issue occurs during
+// deserialization or the YAML input contains invalid fields.
+func UnmarshalPipeline(in []byte) (archer.Manifest, error) {
+	pm := PipelineManifest{}
+	err := yaml.Unmarshal(in, &pm)
+	if err != nil {
+		return nil, err
+	}
+
+	var version PipelineSchemaMajorVersion
+	if version, err = validateVersion(&pm); err != nil {
+		return nil, err
+	}
+
+	// TODO: #221 Do more validations
+	switch version {
+	case Ver1:
+		return &pm, nil
+	}
+	// we should never reach here, this is just to make the compiler happy
+	return nil, errors.New("unexpected error occurs while unmarshalling pipeline.yml")
```

Comment:
I think if we inline the `validateVersion` function instead we can get rid of this comment.

Reply:
For now yes that will do the job but will not completely solve the problem because in the future we will most likely need to do (slightly) different validations per schema version, so this switch statement will still be there. I'm ok with removing it for now and add it later when the needs come. Up to you :)